### PR TITLE
SALTO-3109 - Salesforce: validate account settings

### PR DESCRIPTION
**Pull request by tomsellek**
Add a change validator for `AccountSettings.enableAccountOwnerReport`. This field may only exist if the global account sharing policy is 'Private'

---

The actual value of the global account sharing setting is 'Read', 'Edit', etc. It is never actually set to 'Private' in the API data, but rather translated to other values (including 'Private') in the Salesforce UI.

---
_Release Notes_: 
Salesforce: Improved pre-deploy validation of account settings.

---
_User Notifications_: 
N/A

### CODE REVIEW

File: packages/salesforce-adapter/src/change_validators/account_settings.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    return []
+  }
+
+  const changedInstances = changes
+    .filter(isAdditionOrModificationChange)
+    .map(getChangeData)
+    .filter(isInstanceElement)
+
+  return awu(changedInstances)
+    .filter(isInstanceOfType('AccountSettings'))
```
Comment (Uri-Bar) [inline packages/salesforce-adapter/src/change_validators/account_settings.ts:58]: why did you split it into two part? also, `is instanceOfType` is checking if the element is instance so the first `is instanceElement` is redundant

File: packages/salesforce-adapter/src/change_validators/account_settings.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    elemID: instance.elemID,
+    severity: 'Error',
+    message: 'You cannot set a value for AccountSettings.enableAccountOwnerReport unless your organization-wide sharing access level for Accounts is set to Private',
+    detailedMessage: `In ${instance.elemID.getFullName()}, enableAccountOwnerReport is set to '${instance.value.enableAccountOwnerReport}' but the organization-wide sharing access level for Accounts is not set to 'Private'.`,
+  }
+)
+
+const changeValidator = (client: SalesforceClient): ChangeValidator => async changes => {
+  const query = await buildSelectQueries('Organization', ['DefaultAccountAccess'])
+  const queryResult = await queryClient(client, query)
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/change_validators/account_settings.ts:44]: Just double checking, we don't fetch this information at all right?
@adamro what do you suggest we do here? I know you are quite opinionated about change validators using the client.

File: packages/salesforce-adapter/src/change_validators/account_settings.ts (hunk 1 above)
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/change_validators/account_settings.ts:58]: I think we should extract AccountSettings to the constants file

**Review by amitgoldarb — COMMENTED** [review]
Added some small comments.
Looking good!

File: packages/salesforce-adapter/src/change_validators/account_settings.ts (hunk 2 above)
Comment (adamro) [reply packages/salesforce-adapter/src/change_validators/account_settings.ts:44]: I think we should avoid it when we can.
If we can fetch this information I think that's preferred. If it's interesting to the user then a normal fetch and if not then state-only and use the information from there.

Comment (tomsellek) [reply packages/salesforce-adapter/src/change_validators/account_settings.ts:44]: @tamtamirr this is relevant to our recent discussion. Would you mind talking this through with @adamro instead of me trying to convey what you told me?
יאללה מכות!

Comment (tamtamirr) [reply packages/salesforce-adapter/src/change_validators/account_settings.ts:44]: Me and @adamro already discussed [this](https://salto-io.atlassian.net/browse/SALTO-3386) ticket before. My personal opinion here: Unless we define properly fetch of org settings which means:
- What are the things we consider as org settings? (I can already name a few options: Settings, InstalledPackages, Org Users, And now the `Organization.DefaultAccountAccess` as-well)
- Have a proper flow that we fetch them all from (Not gonna dive into implementation details, but it could be as a filter or part of the fetch itself).

Since we don't have the ability to fetch the settings only (Not through the SAAS as-well), and we still need to define how we store them (Do we want to store them as hidden values? Do we want a dedicated structure?) I think for now this logic should stay in the `deploy` process.

TLDR; In the long term I agree with @adamro that this should be handled in fetch (This could give us other abilities like restricting certain fields to be one of the settings, for example all the `User` fields @tomsellek handles in his recent CV), but our product is not there yet and it feels like out of scope right now. We also need @tomermevorach 's involvement here.

cc. @amitgoldarb

Comment (adamro) [reply packages/salesforce-adapter/src/change_validators/account_settings.ts:44]: What are we worried about?

Whatever is relevant to the user, fetch as we fetch anything else. Whatever is not relevant to the user, fetch as hidden.
Use what we fetched here in the Change Validator.
If we don't have the information, encourage the user to run fetch. 

Supporting partial fetches for these is nice to have IMO and, is clearly not impossible.

Comment (amitgoldarb) [reply packages/salesforce-adapter/src/change_validators/account_settings.ts:44]: @tamtamirr I understand where you're coming from, and I agree that if fetching this and representing it is a large task it might be okay to cut a corner here in this change validator in order to bring value. I _think_ this isn't the case here and we should probably fetch this information and base the CV on fetched nacls.
I also think in the future we will have to update our partial fetch capabilities, for "settings" as well.
I also think this is a slippery slope, if we allow doing this for change validators it will become the norm, we don't want to encourage this type of change validator going forward and prefer to be consistent and base the results on fetched nacls.

For this ticket It I think we will try and not use the client to fetch up to date settings - 
@tomsellek could you please verify how complex it would be to add this information to the state (we can open a ticket to expose it later, this will also save us monitor silencing, it will also limit the task so we don't have to verify we support deploying it, but only fetching it). It should be quite straight forward. After this can you please update this CV to use the fetched state information?

Comment (tomsellek) [reply packages/salesforce-adapter/src/change_validators/account_settings.ts:44]: @amitgoldarb I actually implemented fetching the org-wide settings before my discussion with Tamir, so there's an existing branch that does it - https://github.com/salto-io/salto/pull/3828/files
If that's what we want to do I can rebase this PR on top of the existing fetch one.

Comment (amitgoldarb) [reply packages/salesforce-adapter/src/change_validators/account_settings.ts:44]: I think that is what we would like to do 👍

Comment (tomsellek) [reply packages/salesforce-adapter/src/change_validators/account_settings.ts:44]: @adamro @tamtamirr @amitgoldarb suppose I fetch the global settings instance in a filter (or whatever). By definition it will not change, so won't be part of the changes passed to the CV. As far as I can tell change validators don't receive an ElementsSource.
My question: how can I access elements from a change validator?

Comment (amitgoldarb) [reply packages/salesforce-adapter/src/change_validators/account_settings.ts:44]: I think changeValidators receive changes and an elementSource as parameters.
https://github.com/salto-io/salto/blob/e1b32e19fdf33b47f08a6c773e3d1fedb75d3bbe/packages/adapter-api/src/adapter.ts#L97 - here.
What we pass in the element source is the elements in the nacls - https://github.com/salto-io/salto/blob/e1b32e19fdf33b47f08a6c773e3d1fedb75d3bbe/packages/core/src/core/plan/filter.ts#L204

**Comment by tomsellek** [comment]
Closing this PR to rebase it over https://github.com/salto-io/salto/pull/3829
